### DISCUSSION TITLE
Could SWS be used as k8s ingress controller engine via wrapper och direct integration?

### DISCUSSION THREAD

**Discussion by bjornharrtell** [static-web-server/static-web-server#481] [Ideas]
This is an idea at this point, which I wanted to share and possibly start prototyping on. I'm generally not happy with the k8s ingress alternatives. There are many, most of them rather complex and heavy. An exception is nginx ingress controller. Since I much prefer SWS to nginx I got to thinking SWS could be used in the same way, since SWS nowdays have both rewrite and redirect capabilities.

**Comment by joseluisq** [comment]
It sounds interesting. Honestly, I didn't think about it.
SWS supports rewrites, redirects and other features but I wonder if it should need some additional ones to make it a k8s ingress controller alternative (e.g. load balancing, rate limiting, etc) and the effort necessary.
On the other hand, although SWS supports rewrites and redirects, it was not designed to be a proxy server like nginx and it probably won't be.
However, if you think this can be feasible in some way and have ideas or even a prototype then feel free to share them. Personally, I think it could be not a bad idea, the problem is that I lack some k8s knowledge so I will need help on that front.

**Reply by bjornharrtell** [reply]
Ah I see, it doesn't to reverse proxying. That's probably a blocker for this. And I suppose it was a bit wild idea.. :) But are you sure reverse proxy isn't on the radar? When I saw URL Rewrite and URL Redirect I falsely assumed that it could do that in a reverse proxy like way.

**Reply by joseluisq** [reply]
Well, IMHO If SWS wants to be a "decent" reverse proxy then it should implement several features missing like request forwarding (e.g. GET, POST, etc), backend server forwarding e.g. load balancing, routing by URL patterns or headers (SWS does it in some way), connection management (e.g. keep-alive), SSL proxy, some extra headers manipulation (e.g. X-Forwarded-For, X-Real-IP, etc), maybe some SSL certificates management (e.g. via acme), session persistence (e.g. sticky sessions), then come into play some K8s API integration and other things to consider? just to name a few.

SWS should become more complex quickly. I can also say that I'm not opposed to the idea of becoming a reverse proxy at all someday but this will mean that we will need to redesign SWS and implement a significant number of features if we want to enter the reverse proxy territory.

Regardless, there are some alternatives that maybe you could try together with SWS like https://github.com/sozu-proxy/sozu

**Reply by joseluisq** [reply]
This is just my rough exaggerated estimation. It could maybe be not realistic at all and might turn out to take less effort than that.

**Reply by bjornharrtell** [reply]
Thanks @joseluisq, at least that makes it clear (understandably) it cannot be a short term goal.
